k2/csrc/cuda/debug.h

```diff
@@ -0,0 +1,382 @@
+// k2/csrc/cuda/debug.h
+
+// Copyright (c) 2020, Xiaomi Corporation ( authors: Meixu Song )
+
+// See LICENSE for clarification regarding multiple authors
+
+#ifndef K2_CSRC_CUDA_DEBUG_H_
+#define K2_CSRC_CUDA_DEBUG_H_
+
+/**
+ * To make host compiler preprocessor happy.
+ *
+ * @todo
+ *  Find a way to avoid this and make .h/.cc with cuda code
+ *  could be parsed by host compiler (specially, GNU-gcc).
+ *  (May assgin to nvcc to take control through change cmake
+ *  compiler and options. Then, it need cmake-3.18 `FindCUDAToolkit`
+ *  or other config helpers to make things easy.)
+ */
+#include <cuda_runtime.h>
+#include <cuda_runtime_api.h>
+#include <device_launch_parameters.h>
+#include <driver_types.h>
+
+#include <stdio.h>
+#include <assert.h>
+
+#include "k2/csrc/cuda/arch.h"
+
+namespace k2 {
+
+/**
+ * @brief Define a macro `K2_MAKE_ERROR_CHECK` that turn on the macros bellow
+ *        to make errors get checked, if NDEBUG is undefined.
+ *
+ * @details If in NDEBUG mode, all macros here would do nothing.
+ */
+#if (!defined(NDEBUG) && !defined(K2_MAKE_ERROR_CHECK))
+#define K2_MAKE_ERROR_CHECK
+#endif
+
+/**
+ * @brief A static assertion
+ *
+ * @param[in] exp the compile-time boolean expression that must be true
+ * @param[in] msg an error message if exp is false
+ *
+ * @note `static_assert` is supported by both of host and device.
+ *
+ * @code{.cpp}
+ * K2_STATIC_ASSERT(DEFINED_SHAPE % DEFINED_X == 0);
+ * @endcode
+ */
+#ifndef K2_STATIC_ASSERT
+#define K2_STATIC_ASSERT(exp, msg) static_assert(exp, msg)
```

Comment:
`static_assert` is part of C++11 standard. What is the reason to add another indirection here?

Reply:
agreed, I guess we can just use static_assert.